### PR TITLE
Resource Isolation Feature Checkpoint 2: Selection of compute node considering resource isolation group

### PR DESCRIPTION
**Pull request by ctbrennan**
## What I'm doing:
* When compute nodes have a resource group set, have FE start to manage its own mappings from tablet to compute node, one distinct mapping per resource isolation group. This means delegating to an internal data structure, the singleton `TabletComputeNodeMapper` and bypassing StarOS/StarMgr. This is a significant change to the way StarRocks functions. 
* Whenever there's node selection, return compute nodes which have the same resource isolation group as the current FE
* Improve CN backup selection behavior that kicks in when there's some CN removed or unhealthy: instead of using a single CN as backup for the unhealthy CN, evenly distribute the load originally meant for the unhealthy CN to remaining CN. This is accomplished using consistent hashing with vnodes. 
## Why I'm doing:

* This is part of the implementation of the [Resource Isolation for Shared Data Mode feature](https://docs.google.com/document/d/1oZx-h9bcf7pW_xQSwQyh5ciGjFMxprIrqItPBsKV5G8/edit#heading=h.6jjpflfxnimt)
* This preps the "multiple replicas for shared data mode" feature

Follow up PRs to come which will
* Add the resource isolation group labels to CACHE SELECT statements and implementation of warming up multiple replicas
* Start returning resource isolation groups with `SHOW FRONTENDS` and `SHOW COMPUTE NODES`
See other [milestones](https://docs.google.com/document/d/1Wv1R20gd6DNr5r3dMJQWowQjx7i8UQsHD8s49yk0ei4/edit#heading=h.wgshf257oc6o)
* Start leveraging our new mappings to enable multiple replicas in shared data mode, details to be determined. 
Fixes #issue https://jira.pinadmin.com/browse/RTA-6269

## Testing
Testing ongoing in dev7, asking for review in parallel

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        }
+        map.tabletToComputeNodeId.removeNode(computeNodeId);
+        if (!map.tracksSomeComputeNode()) {
+            resourceIsolationGroupToTabletMapping.remove(resourceIsolationGroup);
+        }
+    }
+
+    public void modifyComputeNode(Long computeNodeId,
+                                  String oldResourceIsolationGroup, String newResourceIsolationGroup) {
+        writeLock.lock();
```
Comment (charW) [inline fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java:152]: before acquiring lock, check if old equals to new resource isolation group, and simply return if so?

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 2]
```diff
@@ -124,6 +129,16 @@ public SystemInfoService() {
...
         nodeSelector = new NodeSelector(this);
+        tabletComputeNodeMapper = new TabletComputeNodeMapper();
+    }
+
+    public boolean shouldUseInternalTabletToCnMapper() {
+        // We prefer to use the TabletComputeNodeMapper rather than delegating to StarOS/StarMgr for tablet->CN mappings
+        // if and only if we're using resource isolation groups.
+        return tabletComputeNodeMapper.numResourceIsolationGroups() > 1;
+    }
+    public TabletComputeNodeMapper internalTabletMapper() {
```
Comment (charW) [inline fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:140]: nit: space between two functions

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 3]
```diff
@@ -124,6 +129,16 @@ public SystemInfoService() {
         pathHashToDishInfoRef = ImmutableMap.of();
 
         nodeSelector = new NodeSelector(this);
+        tabletComputeNodeMapper = new TabletComputeNodeMapper();
+    }
+
+    public boolean shouldUseInternalTabletToCnMapper() {
+        // We prefer to use the TabletComputeNodeMapper rather than delegating to StarOS/StarMgr for tablet->CN mappings
+        // if and only if we're using resource isolation groups.
+        return tabletComputeNodeMapper.numResourceIsolationGroups() > 1;
```
Comment (charW) [inline fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:138]: does this mean that if say we have two non-default groups "group1" and "group2", and all compute nodes in "group2" are dropped for some reason (maybe a bug), then this function will evaluate to false, and all the resource isolation mechanisms will be bypassed entirely, and FEs in "group2" could use the compute nodes in "group1"?

File: fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java (hunk 1 above)
Comment (ctbrennan) [reply fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java:175]: done

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java (hunk 3 above)
Comment (charW) [reply fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java]: Thanks, looks good! I would also add a test case specifically to test this behavior (e.g. have just one group that is a non-default, and make sure the resource isolation mechanism is still used).

**Review by charW — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/lake/qe/scheduler/DefaultSharedDataWorkerProvider.java [hunk 4]
```diff
@@ -225,6 +229,36 @@ public long selectBackupWorker(long workerId) {
...
+                    return -1;
+                }
+                for (Long possibleBackup : cnIdsOrderedByPreference) {
+                    if (possibleBackup != workerId && availableID2ComputeNode.containsKey(possibleBackup) &&
+                            !SimpleScheduler.isInBlocklist(possibleBackup)) {
+                        return possibleBackup;
+                    }
+                }
+                LOG.warn(String.format("Tried to use internal tablet to CN mapper but it failed to find an available" +
+                        " cn for the given tablet %d", tabletId.get()));
```
[the remark this answers is not in the document]
Comment (ctbrennan) [reply fe/fe-core/src/main/java/com/starrocks/lake/qe/scheduler/DefaultSharedDataWorkerProvider.java:258]: I was thinking we could fall back to the other way of choosing a backup but you're probably right that we should fail loudly.

File: fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java [hunk 5]
```diff
@@ -128,6 +129,18 @@ public List<Long> getAllComputeNodeIds(long warehouseId) {
     }
 
     private List<Long> getAllComputeNodeIds(long warehouseId, long workerGroupId) {
+        // If we're using resource isolation groups, we bypass the call to StarOS/StarMgr
+        SystemInfoService systemInfoService = GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo();
+        if (systemInfoService.shouldUseInternalTabletToCnMapper()) {
+            if (warehouseId != DEFAULT_WAREHOUSE_ID || workerGroupId != StarOSAgent.DEFAULT_WORKER_GROUP_ID) {
```
[the remark this answers is not in the document]
Comment (ctbrennan) [reply fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java:135]: Warehouse id is a StarOS concept. The multiple-warehouse feature is the way that the enterprise version implements the same thing we're implementing -- physical resource isolation. See this slack thread: https://starrocks.slack.com/archives/C02FACZSNJV/p1719365538053789

In the open source version we only have the default_warehouse. Basically I don't want us to use multiple physical resource isolation features (ours and the enterprise version) at the same time

File: fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java [hunk 6]
```diff
@@ -189,27 +190,28 @@ public Long getComputeNodeId(String warehouseName, LakeTablet tablet) {
...
-        } catch (StarClientException e) {
+        Set<Long> ids = getAllComputeNodeIdsAssignToTablet(warehouse.getId(), tablet);
+        if (ids != null && !ids.isEmpty()) {
+            return ids.iterator().next();
+        } else {
             return null;
         }
     }
 
     public Set<Long> getAllComputeNodeIdsAssignToTablet(Long warehouseId, LakeTablet tablet) {
```
[the remark this answers is not in the document]
Comment (ctbrennan) [reply fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java:201]: I think there's value in keeping the same names  as in the main starrocks repo where possible, so if/when we merge in, it's clearer that the function is the same.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/Utils.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.starrocks.qe.scheduler;
+
+import com.starrocks.thrift.TScanRange;
+
+import java.util.Optional;
+
+public class Utils {
+    public static Optional<Long> maybeGetTabletId(TScanRange scanRange) {
```
[the remark this answers is not in the document]
Comment (ctbrennan) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/Utils.java:21]: No, not always. The scanRange can be for a non-internal table, in which case there's no way to get a tablet id.

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java (hunk 3 above)
Comment (ctbrennan) [reply fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java]: Good point, done.

**Pull request merged**